### PR TITLE
feat: add puzzle game

### PR DESCRIPTION
**Pull request by jisjoe**
## Status
READY

## Description
The latest updates introduce an engaging new puzzle feature for the upcoming Flutter & Friends 2024 event. This feature includes three question types: matching, single-choice, and fill-in-the-blanks. It also allows for setting time limits and unlocking questions. A leaderboard will be available to track participants' scores. Our proprietary API manages both the quiz questions and the submission process. Additionally, the front end showcases custom Rive animations and audio playback, enhancing the overall user experience.

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: lib/config/http.dart [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import 'dart:io';
+
+class MyHttpOverrides extends HttpOverrides {
+  @override
+  HttpClient createHttpClient(SecurityContext? context) {
+    return super.createHttpClient(context)
+      ..badCertificateCallback =
```
Comment (felangel) [inline lib/config/http.dart:7]: Why is this needed? I don't think we should disable certificate checks

File: lib/puzzles/cubit/puzzles_cubit.dart [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+import 'package:puzzle_repository/puzzle_repository.dart';
+
+part 'puzzles_state.dart';
+
+class PuzzlesCubit extends HydratedCubit<PuzzlesState> {
+  PuzzlesCubit({PuzzleRepository? puzzleRepository})
+      : _puzzleRepository = puzzleRepository ??
+            PuzzleRepository(
+              puzzleApiClient: client.ArkrootPuzzleApiClient(
+                apiKey: '102934e85a7cec4769d8c08207f025e0f1e252bd',
```
Comment (felangel) [inline lib/puzzles/cubit/puzzles_cubit.dart:14]: Did you intend for the API key to be included in the source like this?

File: lib/puzzles/cubit/puzzles_cubit.dart [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+          nextLeaderboardPage: state.nextLeaderboardPage + 1,
+        ),
+      );
+    } catch (_) {
+      emit(state.copyWith(leaderboardFetchingStatus: FetchStatus.failed));
+    }
+  }
+
+  @override
+  PuzzlesState? fromJson(Map<String, dynamic> json) {
```
Comment (felangel) [inline lib/puzzles/cubit/puzzles_cubit.dart:154]: Thoughts on using `@JsonSerializable` instead of manually implementing it by hand?

File: lib/puzzles/widgets/puzzle_listener.dart [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import 'package:flutter_and_friends/puzzles/cubit/puzzles_cubit.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+class PuzzleListener extends StatelessWidget {
+  const PuzzleListener({required this.child, super.key});
+  final Widget child;
+
+  @override
+  Widget build(BuildContext context) {
+    return MultiBlocListener(
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_listener.dart:11]: Can we remove the `MultiBlocListener` since there's only a single listener being used here?

**Comment by felangel** [comment]
Hi @jisjoe (cc @spydon) 👋 

Thanks for opening the PR and adding the puzzle! I just took a quick look and left some comments on the source code, let me know what you think. I also checked out the code and tried running it locally and noticed some issues:

1. I wasn't able to view the leaderboard
2. I wasn't able to play any of the games or answer any of the questions

Overall, I don't feel comfortable merging the code as is because the puzzle integrates with a backend which I don't have access to and since I'm responsible for the quality and stability of the app, I don't feel comfortable merging these changes. If the puzzles don't work or the leaderboard doesn't load there's not much I can do to fix it and in it's current state I don't feel it enhances the application.

If you'd like to integrate some games, I propose something simpler and preferably without external dependencies. It's also worthwhile noting that the entire app is offline-first (it will function just fine even if you don't have any network connection) the changes in the PR make it so that the app requires a stable network connection.

Let me know what you think, thanks again for your contributions, and apologies for any inconvenience! 💙 🙏

**Comment by jisjoe** [comment]
> Hi @jisjoe (cc @spydon) 👋
> 
> Thanks for opening the PR and adding the puzzle! I just took a quick look and left some comments on the source code, let me know what you think. I also checked out the code and tried running it locally and noticed some issues:
> 
> 1. I wasn't able to view the leaderboard
> 2. I wasn't able to play any of the games or answer any of the questions
> 3. The changes significantly increase the size of the application and require a network connection
> 4. The changes introduce a dependency on an external API which I don't control
> 
> Overall, I don't feel comfortable merging the code as is because the puzzle integrates with a backend which I don't have access to and since I'm responsible for the quality and stability of the app, I don't feel comfortable merging these changes. If the puzzles don't work or the leaderboard doesn't load there's not much I can do to fix it and in it's current state I don't feel it enhances the application.
> 
> If you'd like to integrate some games, I propose something simpler and preferably without external dependencies. It's also worthwhile noting that the entire app is offline-first (it will function just fine even if you don't have any network connection) the changes in the PR make it so that the app requires a stable network connection.
> 
> Let me know what you think, thanks again for your contributions, and apologies for any inconvenience! 💙 🙏

@felangel 

Thanks for the review and sorry for the confusion, there shouldn't be any. Please let me clarify the points. 

-  PR comments -  I will fix them as soon as possible and make another PR.
- I couldn't view the leaderboard - The leaderboard was cleared for the final day, and the quizzes were locked for the "the days", I can open them all and give you access to the database as well. (sent details in email)
- the changes introduce a dependency on an external API which I don't control - I can provide access to the DB that I  choose as AWS since I have expertise on the same, and I have shared the access via your email.
-   noting that the entire app is offline first (it will function just fine even if you don't have any network connection)  - Once the quiz is pulled in the app works offline without breaking as it uses hydrated bloc, but users cannot submit answers, being an online quiz the thought was connectivity will be available. 

Can you please take a look again :)

File: lib/puzzles/cubit/puzzles_cubit.dart (hunk 3 above)
Comment (jisjoe) [reply lib/puzzles/cubit/puzzles_cubit.dart:156]: Nice one i have updated it @felangel

File: pubspec.yaml [hunk 5]
```diff
@@ -20,12 +24,19 @@ dependencies:
   hydrated_bloc: ^9.1.5
   intl: ^0.19.0
   json_annotation: ^4.9.0
+  lottie: ^3.1.2
```
Comment (felangel) [inline pubspec.yaml:27]: Why do we need both lottie and rive?

File: pubspec.yaml [hunk 6]
```diff
@@ -20,12 +24,19 @@ dependencies:
...
+  nested_scroll_views: ^0.0.11
+  particles_flutter: ^0.1.4
```
Comment (felangel) [inline pubspec.yaml:29]: Why are these packages needed?

File: pubspec.yaml [hunk 7]
```diff
@@ -7,7 +7,11 @@ environment:
   sdk: ">=3.5.0 <4.0.0"
 
 dependencies:
+  arkroot_puzzle_api:
+    path: packages/arkroot_puzzle_api
+  audioplayers: ^6.0.0
   bloc: ^8.1.4
+  custom_platform_device_id: ^1.0.8
```
Comment (felangel) [inline pubspec.yaml:14]: why is this needed? Our privacy policy states we don't collect any information.

File: lib/puzzles/cubit/puzzles_cubit.dart [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  final PuzzleRepository _puzzleRepository;
+
+  Future<void> verifyUser({required String username}) async {
+    try {
+      emit(
+        state.copyWith(
+          userVerificationStatus: UserVerificationStatus.verifying,
+        ),
+      );
+      final deviceId = await PlatformDeviceId.getDeviceId;
```
Comment (felangel) [inline lib/puzzles/cubit/puzzles_cubit.dart:24]: This change will require me to change the privacy policy of the app since we currently do no collect any device information. Why is this needed?

File: lib/puzzles/cubit/puzzles_cubit.dart [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    required dynamic answer,
+  }) async {
+    try {
+      emit(
+        state.copyWith(
+          currentPuzzleSubmissionStatus: SubmissionStatus.submitting,
+        ),
+      );
+      final userProgress = await _puzzleRepository.submitAnswer(
+        userId: state.user!.userID,
```
Comment (felangel) [inline lib/puzzles/cubit/puzzles_cubit.dart:86]: Shouldn't we ensure user is not null instead of force unwrapping here?

File: lib/puzzles/cubit/puzzles_cubit.dart [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        answer: answer,
+      );
+      _updatePuzzleProgress(
+        puzzleId: puzzleId,
+        userProgress: userProgress,
+      );
+    } on EmptyData catch (_) {
+      final userProgress = UserProgress(
+        score: 0,
+        userID: state.user!.userID.toString(),
```
Comment (felangel) [inline lib/puzzles/cubit/puzzles_cubit.dart:97]: same here about force unwrapping user.

File: lib/puzzles/cubit/puzzles_cubit.dart [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  @override
+  PuzzlesState? fromJson(Map<String, dynamic> json) {
+    final puzzleState = PuzzlesState.fromJson(json);
+    return PuzzlesState(
+      user: puzzleState.user,
+      userVerificationStatus: puzzleState.userVerificationStatus,
+      puzzles: puzzleState.puzzles,
+      leaderboard: puzzleState.leaderboard,
+      nextLeaderboardPage: puzzleState.nextLeaderboardPage,
+    );
```
Comment (felangel) [inline lib/puzzles/cubit/puzzles_cubit.dart:158]: Why aren't we just returning `PuzzlesState.fromJson(json)`?

File: lib/puzzles/widgets/puzzle_types/puzzle_single_choice.dart [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+class PuzzleSingleChoice extends StatelessWidget {
+  const PuzzleSingleChoice({required this.puzzle, super.key});
+
+  final Puzzle puzzle;
+
+  @override
+  Widget build(BuildContext context) {
+    final foxController = FoxController(audioType: AudioType.singleChoice);
+    String? selectedAnswer;
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_types/puzzle_single_choice.dart:17]: This is state that's being maintained in the `build` method which is generally not something you want to do since this widget can rebuilt at any time and lose it's `selectedAnswer`. This should be refactored into a `StatefulWidget` instead if you want to maintain the `selectedAnswer` as state.

File: pubspec.yaml [hunk 13]
```diff
@@ -40,5 +51,20 @@ flutter:
...
+    - family: Poppins
+      fonts:
+        - asset: assets/fonts/poppins/Poppins-Black.ttf
+        - asset: assets/fonts/poppins/Poppins-Bold.ttf
+        - asset: assets/fonts/poppins/Poppins-ExtraBold.ttf
+        - asset: assets/fonts/poppins/Poppins-Light.ttf
+        - asset: assets/fonts/poppins/Poppins-Medium.ttf
+        - asset: assets/fonts/poppins/Poppins-Regular.ttf
+        - asset: assets/fonts/poppins/Poppins-SemiBold.ttf
+        - asset: assets/fonts/poppins/Poppins-Thin.ttf
```
Comment (felangel) [inline pubspec.yaml:70]: Are all these needed? They will increase the overall bundle size and I prefer to keep the app as light as possible.

File: lib/puzzles/cubit/puzzles_cubit.dart (hunk 8 above)
Comment (jisjoe) [reply lib/puzzles/cubit/puzzles_cubit.dart:24]: Hey @felangel , Since we don't have any additional authentication methods, we're binding the username to the `device ID`. For example, if one user chooses the username `john` and another user also selects the same name, we want to be able to differentiate between them based on their `device ID`.

File: pubspec.yaml (hunk 6 above)
Comment (jisjoe) [reply pubspec.yaml:29]: Screenshot 2024-08-28 at 2 09 15 PM

We are currently using the `particles_flutter` package to display the particle effects in the background. The `nested_scroll_views` package is required for the puzzle screen. The quiz questions are rendered using the designated package, and we need the quiz content card to be scrollable while ensuring that the entire screen remains scrollable as well.

File: pubspec.yaml (hunk 5 above)
Comment (jisjoe) [reply pubspec.yaml:27]: We are using Lottie for animations that don't require custom triggers. I prefer to stick with Lottie for displaying simple animations, rather than switching to Rive.

File: pubspec.yaml (hunk 13 above)
Comment (jisjoe) [reply pubspec.yaml:70]: I will remove the font files.

File: lib/puzzles/cubit/puzzles_cubit.dart (hunk 9 above)
Comment (jisjoe) [reply lib/puzzles/cubit/puzzles_cubit.dart:86]: Hey @felangel, We're performing a check here to determine if the user is null. If the user is null, the UI for submitting quiz answers will not be displayed.

https://github.com/felangel/flutter_and_friends/blob/0cf7373868b4d383c8cbfad3e7d64263f03b4c8e/lib/puzzles/widgets/puzzles_page_content.dart#L46

File: pubspec.yaml (hunk 6 above)
Comment (felangel) [reply pubspec.yaml:29]: I think we can achieve nested scroll views without a package and I’m not sure that the particles are really worth adding a whole new package.

File: lib/puzzles/cubit/puzzles_cubit.dart (hunk 8 above)
Comment (jisjoe) [reply lib/puzzles/cubit/puzzles_cubit.dart]: Hey @felangel The issue with using UUIDs is that they need to be cached. If a user uninstalls the app and later tries to access the puzzle again, they won't be able to continue their progress and will have to start as a new user. Anyways i have updated the code to use uuid.

Comment (felangel) [reply lib/puzzles/cubit/puzzles_cubit.dart]: Yeah that's okay though, I think the uninstallation case is fine since users likely won't be installing and uninstalling the app during the conference. Most people will install it once before or during the conference and then uninstall it after the conference. Thanks for making the change!

File: pubspec.yaml [hunk 14]
```diff
@@ -40,5 +51,8 @@ flutter:
     - assets/activities/
     - assets/speakers/
     - assets/sponsors/
+    - assets/lotties/
+    - assets/rive/
+    - assets/audio/
     - shorebird.yaml
-  uses-material-design: true
+  uses-material-design: true
```
Comment (felangel) [inline pubspec.yaml:58]:
```suggestion
  uses-material-design: true

```

File: pubspec.yaml [hunk 15]
```diff
@@ -20,12 +23,20 @@ dependencies:
...
   pusher_beams: ^1.1.1
+  puzzle_repository:
+    path: packages/puzzle_repository
   restart_app: ^1.2.1
+  rive: ^0.13.5
   shorebird_code_push: ^1.1.4
   url_launcher: ^6.3.0
+  uuid: ^3.0.7
+  very_good_infinite_list: ^0.7.1
+
```
Comment (felangel) [inline pubspec.yaml:39]:
```suggestion
  very_good_infinite_list: ^0.7.1
```

Comment (felangel) [inline pubspec.yaml:39]:
```suggestion
  very_good_infinite_list: ^0.7.1
```

File: pubspec.yaml [hunk 16]
```diff
@@ -20,12 +23,20 @@ dependencies:
...
 
 dev_dependencies:
```
Comment (felangel) [inline pubspec.yaml:41]:
```suggestion
  very_good_infinite_list: ^0.7.1

dev_dependencies:
```

File: lib/settings/views/settings_page.dart [hunk 17]
```diff
@@ -1,4 +1,5 @@
 import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/view/puzzles_page.dart';
```
Comment (felangel) [inline lib/settings/views/settings_page.dart:2]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzle_completion/view/puzzle_completion.dart [hunk 18]
```diff
@@ -0,0 +1,62 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzle_leaderboard/view/leaderboard_page.dart';
```
Comment (felangel) [inline lib/puzzle_completion/view/puzzle_completion.dart:2]:
```suggestion
import 'package:flutter_and_friends/puzzle_leaderboard/puzzle_leaderboard.dart';
```

File: lib/puzzle_completion/view/puzzle_completion.dart [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+import 'package:flutter_and_friends/puzzles/helper/puzzle_helper.dart';
+import 'package:flutter_and_friends/puzzles/widgets/fox_animation.dart';
+import 'package:flutter_and_friends/puzzles/widgets/puzzle_button.dart';
```
Comment (felangel) [inline lib/puzzle_completion/view/puzzle_completion.dart:5]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzle_completion/view/puzzle_completion.dart [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+import 'package:flutter_and_friends/theme/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzle_completion/view/puzzle_completion.dart:6]:
```suggestion
import 'package:flutter_and_friends/theme/theme.dart';
```

File: lib/puzzle_leaderboard/view/leaderboard_page.dart [hunk 21]
```diff
@@ -0,0 +1,112 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzle_leaderboard/puzzle_leaderboard.dart';
+import 'package:flutter_and_friends/puzzles/puzzles.dart';
+import 'package:flutter_and_friends/theme/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzle_leaderboard/view/leaderboard_page.dart:4]:
```suggestion
import 'package:flutter_and_friends/theme/theme.dart';
```

File: lib/puzzle_user/widgets/user_submit_button.dart [hunk 22]
```diff
@@ -0,0 +1,85 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzle_user/puzzle_user.dart';
+import 'package:flutter_and_friends/puzzle_user/widgets/validity_info.dart';
```
Comment (felangel) [inline lib/puzzle_user/widgets/user_submit_button.dart:3]:
```suggestion
```

File: lib/puzzles/view/puzzles_page.dart [hunk 23]
```diff
@@ -0,0 +1,37 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/puzzles.dart';
+import 'package:flutter_and_friends/theme/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/view/puzzles_page.dart:3]:
```suggestion
import 'package:flutter_and_friends/theme/theme.dart';
```

File: lib/puzzles/widgets/puzzle_card.dart [hunk 24]
```diff
@@ -0,0 +1,101 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_card.dart:2]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/puzzle_home.dart [hunk 25]
```diff
@@ -0,0 +1,78 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzle_leaderboard/puzzle_leaderboard.dart';
+import 'package:flutter_and_friends/puzzles/helper/puzzle_helper.dart';
+import 'package:flutter_and_friends/puzzles/widgets/fox_animation.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_home.dart:5]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/puzzle_listener.dart [hunk 26]
```diff
@@ -0,0 +1,19 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/cubit/puzzles_cubit.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_listener.dart:2]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/puzzle_types/puzzle_input.dart [hunk 27]
```diff
@@ -0,0 +1,169 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_and_friends/puzzles/cubit/puzzles_cubit.dart';
+import 'package:flutter_and_friends/puzzles/helper/puzzle_helper.dart';
+import 'package:flutter_and_friends/puzzles/widgets/fox_animation.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_types/puzzle_input.dart:6]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/puzzle_types/puzzle_match.dart [hunk 28]
```diff
@@ -0,0 +1,181 @@
+import 'dart:math';
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/cubit/puzzles_cubit.dart';
+import 'package:flutter_and_friends/puzzles/helper/puzzle_helper.dart';
+import 'package:flutter_and_friends/puzzles/models/models.dart';
+import 'package:flutter_and_friends/puzzles/widgets/fox_animation.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_types/puzzle_match.dart:7]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/puzzle_types/puzzle_single_choice.dart [hunk 29]
```diff
@@ -0,0 +1,64 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/cubit/puzzles_cubit.dart';
+import 'package:flutter_and_friends/puzzles/helper/puzzle_helper.dart';
+import 'package:flutter_and_friends/puzzles/widgets/fox_animation.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzle_types/puzzle_single_choice.dart:5]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/puzzles_page_content.dart [hunk 30]
```diff
@@ -0,0 +1,123 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzle_completion/view/puzzle_completion.dart';
+import 'package:flutter_and_friends/puzzle_user/view/view.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/puzzles_page_content.dart:3]:
```suggestion
import 'package:flutter_and_friends/puzzle_completion/puzzle_completion.dart';
import 'package:flutter_and_friends/puzzle_user/puzzle_user.dart';
```

File: lib/puzzles/widgets/status_widgets/puzzle_completed.dart [hunk 31]
```diff
@@ -0,0 +1,78 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/helper/puzzle_helper.dart';
+import 'package:flutter_and_friends/puzzles/widgets/fox_animation.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/status_widgets/puzzle_completed.dart:4]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/status_widgets/puzzle_locked.dart [hunk 32]
```diff
@@ -0,0 +1,47 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/cubit/puzzles_cubit.dart';
+import 'package:flutter_and_friends/puzzles/widgets/time_counter.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/status_widgets/puzzle_locked.dart:3]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: lib/puzzles/widgets/status_widgets/puzzle_unlocked.dart [hunk 33]
```diff
@@ -0,0 +1,71 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_and_friends/puzzles/widgets/widgets.dart';
```
Comment (felangel) [inline lib/puzzles/widgets/status_widgets/puzzle_unlocked.dart:2]:
```suggestion
import 'package:flutter_and_friends/puzzles/puzzles.dart';
```

File: packages/arkroot_puzzle_api/lib/src/models/puzzle.dart [hunk 34]
```diff
@@ -0,0 +1,65 @@
+import 'package:arkroot_puzzle_api/src/models/user_progress.dart';
```
Comment (felangel) [inline packages/arkroot_puzzle_api/lib/src/models/puzzle.dart:1]:
```suggestion
import 'package:arkroot_puzzle_api/arkroot_puzzle_api.dart';
```

File: packages/puzzle_repository/lib/src/models/puzzle.dart [hunk 35]
```diff
@@ -0,0 +1,101 @@
+import 'dart:async';
+
+import 'package:equatable/equatable.dart';
+import 'package:json_annotation/json_annotation.dart';
+import 'package:ntp/ntp.dart';
+import 'package:puzzle_repository/src/models/user_progress.dart';
```
Comment (felangel) [inline packages/puzzle_repository/lib/src/models/puzzle.dart:6]:
```suggestion
import 'package:puzzle_repository/puzzle_repository.dart';
```

**Comment by felangel** [comment]
Thanks for making all the changes! I'll get this merged and released later today.

**Comment by jisjoe** [comment]
> Thanks for making all the changes! I'll get this merged and released later today.

Hey, @felangel , shall I replace the final questions then?

**Comment by felangel** [comment]
> > Thanks for making all the changes! I'll get this merged and released later today.
> 
> Hey, @felangel, shall I replace the final questions then? They will be locked and only unlocked on September 1.

Sure, do I still use the same API_KEY to build the app?

**Comment by jisjoe** [comment]
> > > Thanks for making all the changes! I'll get this merged and released later today.
> > 
> > 
> > Hey, @felangel, shall I replace the final questions then? They will be locked and only unlocked on September 1.
> 
> Sure, do I still use the same API_KEY to build the app?

Yes @felangel , we are using the same API_KEY to build the app.

**Comment by felangel** [comment]
@jisjoe I rebased and squashed everything in #11

I'm just going to close this PR and leave the changes on the other branch if that's okay with you. I'll build and release off of the branch shortly 👍 

Thanks for all your hard work! 💙 🙏
